osmosis-labs/osmosis-frontend
# 🧹 (Dust to OSMO): Feature implementation

## What is the purpose of the change:

This PR is the v0 for the dust to OSMO feature. 

### Outline for the solution

- add a `Dust to OSMO` button to the portfolio page under the asset table 
- when the user clicks `Swap dust to OSMO`
  - loading spinner is added to the button
  - a swap is generated for every asset, and swaps are executed sequentially
  - user is prompted to sign every tx by the wallet
  - after all swap txs resulted in either success or error remove the loading spinner

### Linear Task

<details>
<summary>Brief for the task</summary>

>Currently, a user who regularly trades on the platform may have a number of small balances remaining in their portfolio due to the values of various tokens being incommensurate. These small remaining balances are known as 'dust' and often sit unused since the value is too low in order to justify transacting due to network fees. Your task is to write a function that converts the dust in a wallet (defined as values less than $0.01USD) to OSMO.
>
> Note that there are no designs for this task. You may either write a well tested function for review, or use your best judgement in adding UX to trigger the function.
>
> The task will involve reading and understanding how balances are fetched in the codebase, and understanding signing broadcasting transactions.

</details>

## Brief Changelog

- add a feature flag 
- add the dust to osmo button
- update `useSwap` to be easier to use programatically
- add `getUserDustAsset()` procedure
- on user action execute all the swaps sequentially either with manual signing or 1click trading
- add react query devtools for debugging 
- add a generic query client for non trpc async handling

## Testing and Verifying

This change has been tested locally by rebuilding the website and verified content and links are expected

- set `NEXT_PUBLIC_SHOW_DUST_TO_OSMO=true` in `packages/web/.env.local` 
- go to portfolio 
- make sure you have some dust in the currently connected wallet
- (1click trading off): click `Dust to OSMO` button
- (1click trading on): click `Dust to OSMO` button

## Backlog 

- set up the flag in launch darkly
- drop the environment variable from `.env` and `env.ts` and hardcode a default value in `web/hooks/use-feature-flags.ts`
- Resolve the  `TODO(Greg)` comments 

## To be discussed
- `TODO(Greg)` comments 
- Do we want an overview, confirmation, and error display user interface?
- Should we batch the swaps into a single tx so only 1 signature is necessary?
- Do we want to reduce dust sweep costs? 

## Demo 
### manual sign
![CleanShot 2024-11-05 at 09 09 36](https://github.com/user-attachments/assets/0ab32c8a-a856-4d8a-840d-0bc7453cdfe2)

### 1click trade
![CleanShot 2024-11-05 at 09 15 59](https://github.com/user-attachments/assets/a0db8a1c-9b7b-4f54-baa1-b2168705c77a)
